homeassistant/components/sonos/__init__.py

```diff
@@ -47,9 +90,103 @@ async def async_setup(hass, config):
+        hosts = config.get(CONF_HOSTS)
+
+        def _discovered_player(soco: SoCo) -> None:
+            """Handle a (re)discovered player."""
+            try:
+                _LOGGER.debug("Reached _discovered_player, soco=%s", soco)
+
+                data = hass.data[DATA_SONOS]
+
+                if soco.uid not in data.discovered:
+                    _LOGGER.debug("Adding new speaker")
+                    speaker_info = soco.get_speaker_info(True)
+                    speaker = SonosSpeaker(hass, soco, speaker_info)
+                    data.discovered[soco.uid] = speaker
+                    speaker.setup()
+                else:
+                    dispatcher_send(hass, f"{SONOS_SEEN}-{soco.uid}", soco)
+
+            except SoCoException as ex:
+                _LOGGER.debug("SoCoException, ex=%s", ex)
+
+        if hosts:
+            for host in hosts:
+                try:
+                    _LOGGER.debug("Testing %s", host)
+                    player = pysonos.SoCo(socket.gethostbyname(host))
+                    if player.is_visible:
+                        # Make sure that the player is available
+                        _ = player.volume
+
+                        _discovered_player(player)
+                except (OSError, SoCoException) as ex:
+                    _LOGGER.debug("Exception %s", ex)
+                    if now is None:
+                        _LOGGER.warning("Failed to initialize '%s'", host)
+
+            _LOGGER.debug("Tested all hosts")
+            hass.data[DATA_SONOS].hosts_heartbeat = hass.helpers.event.call_later(
+                DISCOVERY_INTERVAL.total_seconds(), _discovery
+            )
+        else:
+            _LOGGER.debug("Starting discovery thread")
+            hass.data[DATA_SONOS].discovery_thread = pysonos.discover_thread(
+                _discovered_player,
+                interval=DISCOVERY_INTERVAL.total_seconds(),
+                interface_addr=config.get(CONF_INTERFACE_ADDR),
+            )
+            hass.data[DATA_SONOS].discovery_thread.name = "Sonos-Discovery"
+
+    @callback
+    def _async_signal_update_groups(event):
+        async_dispatcher_send(hass, SONOS_GROUP_UPDATE)
+
+    @callback
+    def start_discovery():
+        _LOGGER.debug("Adding discovery job")
+        hass.async_add_executor_job(_discovery)
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _stop_discovery)
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, _async_signal_update_groups
+        )
+
+    @callback
+    def platform_ready(platform, _):
+        hass.data[DATA_SONOS].platforms_ready.add(platform)
+        if hass.data[DATA_SONOS].platforms_ready == PLATFORMS:
+            start_discovery()
+
+    for platform in PLATFORMS:
+        task = hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+        task.add_done_callback(partial(platform_ready, platform))
```

Comment:
We can await the forwarding with asyncio.gather if we wrap this in a coroutine function that we schedule in a task. See eg:

https://github.com/home-assistant/core/blob/e5e71c2026d304ffd2b9d26c89f2cc28eadb5505/homeassistant/components/ozw/__init__.py#L263-L269

https://github.com/home-assistant/core/blob/e5e71c2026d304ffd2b9d26c89f2cc28eadb5505/homeassistant/components/ozw/__init__.py#L296

Then we can avoid adding a callback to the task which is a low level api.